{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "279de085",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['a001', 'a002', 'a1']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(['a1', 'a001', 'a002', ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "53b6773e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "239119.05109482087"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N_JP0 = 4384  # Japanese population size (inital)\n",
    "r_JP = 0.0129  # Japanese growth rate\n",
    "generation_time = 29\n",
    "T_CH_JP = int(9000 / generation_time)  # time of Han/Japanese split\n",
    "\n",
    "N_JP0 * math.exp(r_JP * T_CH_JP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0058a323",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import math\n",
    "import tskit\n",
    "import tszip\n",
    "import msprime\n",
    "import stdpopsim\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import collections\n",
    "import multiprocessing\n",
    "import itertools\n",
    "from contextlib import redirect_stdout\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "98d64599",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.1.0'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "msprime.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f9a1065a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('0.4.1', '1.1.0', '0.1.2')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tskit.__version__, msprime.__version__, stdpopsim.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d5cf22a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "_species = stdpopsim.get_species(\"HomSap\")\n",
    "_yri_population = stdpopsim.Population(\n",
    "    id=\"YRI\", description=\"1000 Genomes YRI (Yoruba)\"\n",
    ")\n",
    "_ceu_population = stdpopsim.Population(\n",
    "    id=\"CEU\",\n",
    "    description=(\n",
    "        \"1000 Genomes CEU (Utah Residents (CEPH) with Northern and \"\n",
    "        \"Western European Ancestry\"\n",
    "    ),\n",
    ")\n",
    "_chb_population = stdpopsim.Population(\n",
    "    id=\"CHB\", description=\"1000 Genomes CHB (Han Chinese in Beijing, China)\"\n",
    ")\n",
    "\n",
    "def _ooa_4pop():\n",
    "    id = \"OutOfAfrica_4J17\"\n",
    "    description = \"4 population out of Africa\"\n",
    "    long_description = \"\"\"\n",
    "        Demographic model for a four population out-of-Africa history,\n",
    "        taken from Jouganous et al. (2017). Parameter values were taken\n",
    "        from table 4 in the main text. This model was fit based on joint\n",
    "        allele frequecy spectrum (AFS) data from 1000 Genomes exomes\n",
    "        from the YRI, CEU, CHB, and JPT poulation samples. The demography\n",
    "        follows the previous three-populations out-of-Africa models\n",
    "        with an additional population split in Asia leading to the\n",
    "        Japanese (JPT) population. Parameter values were estimated with\n",
    "        the program Moments assuming a mutation rate of 1.44e-8 and a\n",
    "        generation time of 29 years.\n",
    "    \"\"\"\n",
    "    populations = [\n",
    "        _yri_population,\n",
    "        _ceu_population,\n",
    "        _chb_population,\n",
    "        stdpopsim.Population(\n",
    "            id=\"JPT\",\n",
    "            description=\"1000 Genomes JPT (Japanese in Tokyo, Japan)\",\n",
    "            sampling_time=0,\n",
    "        ),\n",
    "    ]\n",
    "    citations = [\n",
    "        stdpopsim.Citation(\n",
    "            author=\"Jouganous et al.\",\n",
    "            year=\"2017\",\n",
    "            doi=\"https://doi.org/10.1534/genetics.117.200493\",\n",
    "            reasons={stdpopsim.CiteReason.DEM_MODEL},\n",
    "        )\n",
    "    ]\n",
    "\n",
    "    generation_time = 29\n",
    "    mutation_rate = 1.44e-8  # from Gravel et al. 2013 (Plos Gen)\n",
    "\n",
    "    # Parameter values from Table 4 (in bold)\n",
    "\n",
    "    # Population sizes\n",
    "    N_A = 11293  # ancestral YRI population size\n",
    "    N_AF = 23721  # YRI population size\n",
    "    N_B = 2831  # B population size (out-of-Africa)\n",
    "    N_EU0 = 2512  # CEU population size (intial)\n",
    "    N_AS0 = 1019  # Asian population size (inital)\n",
    "    N_JP0 = 4384  # Japanese population size (inital)\n",
    "\n",
    "    # growth rates\n",
    "    r_EU = 0.0016  # European growth rate\n",
    "    r_JP = 0.0129  # Japanese growth rate\n",
    "    r_AS = 0.0026  # Asian growth rate\n",
    "\n",
    "    # migration rates\n",
    "    m_AF_B = 16.8e-5  # YRI <--> B (out-of-Africa)\n",
    "    m_AF_EU = 1.14e-5  # YRI <--> CEU\n",
    "    m_AF_AS = 0.56e-5  # YRI <--> CHB\n",
    "    m_EU_AS = 4.75e-5  # European <--> CHB\n",
    "    m_CH_JP = 3.3e-5  # CHBHan <--> JPT\n",
    "\n",
    "    # epochs (specified in years, converted to generations)\n",
    "    T_AF = int(357000 / generation_time)  # time ancestral to African\n",
    "    T_B = int(119000 / generation_time)  # time of Out of Africa split\n",
    "    T_EU_AS = int(46000 / generation_time)  # time of Europe/Asia split\n",
    "    T_CH_JP = int(9000 / generation_time)  # time of Han/Japanese split\n",
    "\n",
    "    # also\n",
    "    r_AF = 0  # YRI and B growth rate\n",
    "\n",
    "    # A and YRI\n",
    "    pop0 = msprime.PopulationConfiguration(\n",
    "        initial_size=N_AF * math.exp(r_AF * T_AF),\n",
    "        growth_rate=r_AF,\n",
    "        metadata=populations[0].asdict(),\n",
    "    )\n",
    "\n",
    "    # B and CEU\n",
    "    pop1 = msprime.PopulationConfiguration(\n",
    "        initial_size=N_EU0 * math.exp(r_EU * T_EU_AS),\n",
    "        growth_rate=r_EU,\n",
    "        metadata=populations[1].asdict(),\n",
    "    )\n",
    "\n",
    "    # Asian and CHB\n",
    "    pop2 = msprime.PopulationConfiguration(\n",
    "        initial_size=N_AS0 * math.exp(r_AS * T_EU_AS),\n",
    "        growth_rate=r_AS,\n",
    "        metadata=populations[2].asdict(),\n",
    "    )\n",
    "\n",
    "    # JPT\n",
    "    pop3 = msprime.PopulationConfiguration(\n",
    "        initial_size=N_JP0 * math.exp(r_JP * T_CH_JP),\n",
    "        growth_rate=r_JP,\n",
    "        metadata=populations[3].asdict(),\n",
    "    )\n",
    "\n",
    "    population_configurations = [pop0, pop1, pop2, pop3]\n",
    "\n",
    "    migration_matrix = [\n",
    "        [0, m_AF_EU, m_AF_AS, 0],\n",
    "        [m_AF_EU, 0, m_EU_AS, 0],\n",
    "        [m_AF_AS, m_EU_AS, 0, m_CH_JP],\n",
    "        [0, 0, m_CH_JP, 0],\n",
    "    ]\n",
    "\n",
    "    # Demographic events from most recent -> oldest\n",
    "\n",
    "    # JPT splits from CHB\n",
    "    JPT_event = [\n",
    "        # seg migration rate JPT <--> CHB to zero\n",
    "        msprime.MigrationRateChange(time=T_CH_JP, rate=0.0, matrix_index=(2, 3)),\n",
    "        msprime.MigrationRateChange(time=T_CH_JP, rate=0.0, matrix_index=(3, 2)),\n",
    "        # JPT splits from CHB\n",
    "        msprime.MassMigration(time=T_CH_JP, source=3, destination=2, proportion=1.0),\n",
    "    ]\n",
    "\n",
    "    # CEU and CHB merge into B\n",
    "    AS_event = [\n",
    "        # set migration rates to zero\n",
    "        msprime.MigrationRateChange(time=T_EU_AS, rate=0.0),\n",
    "        # CHB splits from CEU\n",
    "        msprime.MassMigration(time=T_EU_AS, source=2, destination=1, proportion=1.0),\n",
    "        # change size of B to N_B\n",
    "        # set growth rates in B to zero\n",
    "        msprime.PopulationParametersChange(\n",
    "            time=T_EU_AS, initial_size=N_B, growth_rate=0.0, population_id=1\n",
    "        ),\n",
    "        # set up migration between YRI and B\n",
    "        msprime.MigrationRateChange(time=T_EU_AS, rate=m_AF_B, matrix_index=(0, 1)),\n",
    "        msprime.MigrationRateChange(time=T_EU_AS, rate=m_AF_B, matrix_index=(1, 0)),\n",
    "    ]\n",
    "\n",
    "    # Out-of-Africa event\n",
    "    EU_event = [\n",
    "        # B splits from YRI\n",
    "        msprime.MigrationRateChange(time=T_B, rate=0.0),\n",
    "        msprime.MassMigration(time=T_B, source=1, destination=0, proportion=1.0),\n",
    "    ]\n",
    "\n",
    "    # Change size of YRI/African population\n",
    "    AF_event = [\n",
    "        msprime.PopulationParametersChange(time=T_AF, initial_size=N_A, population_id=0)\n",
    "    ]\n",
    "\n",
    "    demographic_events = JPT_event + AS_event + EU_event + AF_event\n",
    "\n",
    "    return stdpopsim.DemographicModel(\n",
    "        id=id,\n",
    "        description=description,\n",
    "        long_description=long_description,\n",
    "        populations=populations,\n",
    "        citations=citations,\n",
    "        generation_time=generation_time,\n",
    "        #mutation_rate=mutation_rate,\n",
    "        population_configurations=population_configurations,\n",
    "        migration_matrix=migration_matrix,\n",
    "        demographic_events=demographic_events,\n",
    "    )\n",
    "\n",
    "\n",
    "_species.add_demographic_model(_ooa_4pop())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f452538",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "62de0ac0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DemographyDebugger\n",
      "╠════════════════════════════════╗\n",
      "║ Epoch[0]: [0, 310) generations ║\n",
      "╠════════════════════════════════╝\n",
      "╟    Populations (total=4 active=4)\n",
      "║    ┌───────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │      start│       end│growth_rate  │  pop_0   │  pop_1   │  pop_2   │  pop_3  │\n",
      "║    ├───────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│    23721.0│   23721.0│ 0           │    0     │ 1.14e-05 │ 5.6e-06  │    0    │\n",
      "║    │  pop_1│    31775.0│   19349.7│ 0.0016      │ 1.14e-05 │    0     │ 4.75e-05 │    0    │\n",
      "║    │  pop_2│    62955.1│   28118.3│ 0.0026      │ 5.6e-06  │ 4.75e-05 │    0     │ 3.3e-05 │\n",
      "║    │  pop_3│   239119.1│    4384.0│ 0.0129      │    0     │    0     │ 3.3e-05  │    0    │\n",
      "║    └───────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 310\n",
      "║    ┌────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │  time│type            │parameters         │effect                                  │\n",
      "║    ├────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │   310│Migration rate  │source=2, dest=3,  │Backwards-time migration rate from 2    │\n",
      "║    │      │change          │rate=0.0           │to 3 → 0.0                              │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │   310│Migration rate  │source=3, dest=2,  │Backwards-time migration rate from 3    │\n",
      "║    │      │change          │rate=0.0           │to 2 → 0.0                              │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │   310│Mass Migration  │source=3, dest=2,  │All lineages currently in population 3  │\n",
      "║    │      │                │proportion=1       │move to 2 (equivalent to individuals    │\n",
      "║    │      │                │                   │migrating from 2 to 3 forwards in       │\n",
      "║    │      │                │                   │time)                                   │\n",
      "║    └────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠═══════════════════════════════════════╗\n",
      "║ Epoch[1]: [310, 1.59e+03) generations ║\n",
      "╠═══════════════════════════════════════╝\n",
      "╟    Populations (total=4 active=4)\n",
      "║    ┌────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │     start│       end│growth_rate  │  pop_0   │  pop_1   │  pop_2   │ pop_3 │\n",
      "║    ├────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│   23721.0│   23721.0│ 0           │    0     │ 1.14e-05 │ 5.6e-06  │   0   │\n",
      "║    │  pop_1│   19349.7│    2512.0│ 0.0016      │ 1.14e-05 │    0     │ 4.75e-05 │   0   │\n",
      "║    │  pop_2│   28118.3│    1019.0│ 0.0026      │ 5.6e-06  │ 4.75e-05 │    0     │   0   │\n",
      "║    │  pop_3│    4384.0│       0.0│ 0.0129      │    0     │    0     │    0     │   0   │\n",
      "║    └────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 1.59e+03\n",
      "║    ┌──────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │  time│type            │parameters           │effect                                  │\n",
      "║    ├──────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  1586│Migration rate  │source=-1, dest=-1,  │Backwards-time migration rate for all   │\n",
      "║    │      │change          │rate=0.0             │populations → 0.0                       │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  1586│Mass Migration  │source=2, dest=1,    │All lineages currently in population 2  │\n",
      "║    │      │                │proportion=1         │move to 1 (equivalent to individuals    │\n",
      "║    │      │                │                     │migrating from 1 to 2 forwards in       │\n",
      "║    │      │                │                     │time)                                   │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  1586│Population      │population=1,        │initial_size → 2.8e+03 and growth_rate  │\n",
      "║    │      │parameter       │initial_size=2831,   │→ 0 for population 1                    │\n",
      "║    │      │change          │growth_rate=0.0      │                                        │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  1586│Migration rate  │source=0, dest=1,    │Backwards-time migration rate from 0    │\n",
      "║    │      │change          │rate=0.000168        │to 1 → 0.000168                         │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  1586│Migration rate  │source=1, dest=0,    │Backwards-time migration rate from 1    │\n",
      "║    │      │change          │rate=0.000168        │to 0 → 0.000168                         │\n",
      "║    └──────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠═══════════════════════════════════════════╗\n",
      "║ Epoch[2]: [1.59e+03, 4.1e+03) generations ║\n",
      "╠═══════════════════════════════════════════╝\n",
      "╟    Populations (total=4 active=4)\n",
      "║    ┌─────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │     start│       end│growth_rate  │  pop_0   │  pop_1   │ pop_2 │ pop_3 │\n",
      "║    ├─────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│   23721.0│   23721.0│ 0           │    0     │ 0.000168 │   0   │   0   │\n",
      "║    │  pop_1│    2831.0│    2831.0│ 0           │ 0.000168 │    0     │   0   │   0   │\n",
      "║    │  pop_2│    1019.0│       1.5│ 0.0026      │    0     │    0     │   0   │   0   │\n",
      "║    │  pop_3│       0.0│       0.0│ 0.0129      │    0     │    0     │   0   │   0   │\n",
      "║    └─────────────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 4.1e+03\n",
      "║    ┌──────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │  time│type            │parameters           │effect                                  │\n",
      "║    ├──────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  4103│Migration rate  │source=-1, dest=-1,  │Backwards-time migration rate for all   │\n",
      "║    │      │change          │rate=0.0             │populations → 0.0                       │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  4103│Mass Migration  │source=1, dest=0,    │All lineages currently in population 1  │\n",
      "║    │      │                │proportion=1         │move to 0 (equivalent to individuals    │\n",
      "║    │      │                │                     │migrating from 0 to 1 forwards in       │\n",
      "║    │      │                │                     │time)                                   │\n",
      "║    └──────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠═══════════════════════════════════════════╗\n",
      "║ Epoch[3]: [4.1e+03, 1.23e+04) generations ║\n",
      "╠═══════════════════════════════════════════╝\n",
      "╟    Populations (total=4 active=4)\n",
      "║    ┌───────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │     start│       end│growth_rate  │ pop_0 │ pop_1 │ pop_2 │ pop_3 │\n",
      "║    ├───────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│   23721.0│   23721.0│ 0           │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_1│    2831.0│    2831.0│ 0           │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_2│       1.5│       0.0│ 0.0026      │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_3│       0.0│       0.0│ 0.0129      │   0   │   0   │   0   │   0   │\n",
      "║    └───────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 1.23e+04\n",
      "║    ┌─────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       time│type        │parameters          │effect                                 │\n",
      "║    ├─────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  1.231e+04│Population  │population=0,       │initial_size → 1.1e+04 for population  │\n",
      "║    │           │parameter   │initial_size=11293  │0                                      │\n",
      "║    │           │change      │                    │                                       │\n",
      "║    └─────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠═══════════════════════════════════════╗\n",
      "║ Epoch[4]: [1.23e+04, inf) generations ║\n",
      "╠═══════════════════════════════════════╝\n",
      "╟    Populations (total=4 active=4)\n",
      "║    ┌───────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │     start│       end│growth_rate  │ pop_0 │ pop_1 │ pop_2 │ pop_3 │\n",
      "║    ├───────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│   11293.0│   11293.0│ 0           │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_1│    2831.0│    2831.0│ 0           │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_2│       0.0│       0.0│ 0.0026      │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_3│       0.0│       0.0│ 0.0129      │   0   │   0   │   0   │   0   │\n",
      "║    └───────────────────────────────────────────────────────────────────────────┘\n"
     ]
    }
   ],
   "source": [
    "_species.get_demographic_model(\"OutOfAfrica_4J17\").debug()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "93b9ba9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _ooa_4popADMIXTURE():\n",
    "    id = \"OutOfAfrica_4J17_ADMIXTURE\"\n",
    "    description = \"4 population out of Africa with admixture\"\n",
    "    long_description = \"\"\"\n",
    "        Demographic model for a four population out-of-Africa history,\n",
    "        taken from Jouganous et al. (2017), \n",
    "        with admixture and recent population growth added.  \n",
    "    \"\"\"\n",
    "    populations = [\n",
    "        _yri_population,\n",
    "        _ceu_population,\n",
    "        _chb_population,\n",
    "        stdpopsim.Population(\n",
    "            id=\"JPT\",\n",
    "            description=\"1000 Genomes JPT (Japanese in Tokyo, Japan)\",\n",
    "            sampling_time=0,\n",
    "        ),\n",
    "        stdpopsim.Population(\n",
    "            id=\"ADMIX\",\n",
    "            description=\"Admixed population\",\n",
    "            sampling_time=0,\n",
    "        ),        \n",
    "    ]\n",
    "    citations = [\n",
    "        stdpopsim.Citation(\n",
    "            author=\"Jouganous et al.\",\n",
    "            year=\"2017\",\n",
    "            doi=\"https://doi.org/10.1534/genetics.117.200493\",\n",
    "            reasons={stdpopsim.CiteReason.DEM_MODEL},\n",
    "        )\n",
    "    ]\n",
    "\n",
    "    generation_time = 29\n",
    "    # Parameter values from Table 4 (in bold)\n",
    "\n",
    "    # Population sizes\n",
    "    N_A = 11293  # ancestral YRI population size\n",
    "    N_AF = 23721  # YRI population size\n",
    "    N_B = 2831  # B population size (out-of-Africa)\n",
    "    N_EU0 = 2512  # CEU population size (intial)\n",
    "    N_AS0 = 1019  # Asian population size (inital)\n",
    "    N_JP0 = 4384  # Japanese population size (inital)\n",
    "\n",
    "    # growth rates\n",
    "    r_EU = 0.0016  # European growth rate\n",
    "    r_JP = 0.0129  # Japanese growth rate\n",
    "    r_AS = 0.0026  # Asian growth rate\n",
    "    \n",
    "    # added recent growth\n",
    "    recent_AF = 0.1438809454231192\n",
    "    recent_EU = 0.11464909373821389\n",
    "    recent_AS = 0.04627484317475039\n",
    "\n",
    "    # migration rates\n",
    "    m_AF_B = 16.8e-5  # YRI <--> B (out-of-Africa)\n",
    "    m_AF_EU = 1.14e-5  # YRI <--> CEU\n",
    "    m_AF_AS = 0.56e-5  # YRI <--> CHB\n",
    "    m_EU_AS = 4.75e-5  # European <--> CHB\n",
    "    m_CH_JP = 3.3e-5  # CHBHan <--> JPT\n",
    "\n",
    "    # epochs (specified in years, converted to generations)\n",
    "    T_AF = int(357000 / generation_time)  # time ancestral to African\n",
    "    T_B = int(119000 / generation_time)  # time of Out of Africa split\n",
    "    T_EU_AS = int(46000 / generation_time)  # time of Europe/Asia split\n",
    "    T_CH_JP = int(9000 / generation_time)  # time of Han/Japanese split\n",
    "\n",
    "    # also\n",
    "    r_AF = 0  # YRI and B growth rate\n",
    "\n",
    "    # A and YRI\n",
    "    pop0 = msprime.PopulationConfiguration(\n",
    "        #initial_size=N_AF * math.exp(r_AF * T_AF),\n",
    "        #growth_rate=r_AF,\n",
    "        initial_size=100000,\n",
    "        growth_rate=recent_AF,\n",
    "        metadata=populations[0].asdict(),\n",
    "    )\n",
    "\n",
    "    # B and CEU\n",
    "    pop1 = msprime.PopulationConfiguration(\n",
    "        #initial_size=N_EU0 * math.exp(r_EU * T_EU_AS),\n",
    "        #growth_rate=r_EU,\n",
    "        initial_size=100000,\n",
    "        growth_rate=recent_AF,\n",
    "        metadata=populations[1].asdict(),\n",
    "    )\n",
    "\n",
    "    # Asian and CHB\n",
    "    pop2 = msprime.PopulationConfiguration(\n",
    "        #initial_size=N_AS0 * math.exp(r_AS * T_EU_AS),\n",
    "        #growth_rate=r_AS,\n",
    "        initial_size=100000,\n",
    "        growth_rate=recent_AS,\n",
    "        metadata=populations[2].asdict(),\n",
    "    )\n",
    "\n",
    "    # JPT\n",
    "    pop3 = msprime.PopulationConfiguration(\n",
    "        initial_size=N_JP0 * math.exp(r_JP * T_CH_JP),\n",
    "        growth_rate=r_JP,\n",
    "        metadata=populations[3].asdict(),\n",
    "    )\n",
    "    \n",
    "    # ADMIX\n",
    "    pop4 = msprime.PopulationConfiguration(\n",
    "        initial_size=50000 * math.exp(0.05 * 12),\n",
    "        growth_rate=0.05,\n",
    "        metadata=populations[4].asdict(),\n",
    "    )\n",
    "\n",
    "    population_configurations = [pop0, pop1, pop2, pop3, pop4]\n",
    "\n",
    "    migration_matrix = [\n",
    "        [0, m_AF_EU, m_AF_AS, 0, 0],\n",
    "        [m_AF_EU, 0, m_EU_AS, 0, 0],\n",
    "        [m_AF_AS, m_EU_AS, 0, m_CH_JP, 0],\n",
    "        [0, 0, m_CH_JP, 0, 0],\n",
    "        [0,0,0,0,0]\n",
    "    ]\n",
    "\n",
    "    # Demographic events from most recent -> oldest\n",
    "    \n",
    "    # growth\n",
    "    stop_recent_growth = [\n",
    "        msprime.PopulationParametersChange(\n",
    "            population_id=0, time=10, growth_rate=r_AF\n",
    "        ),\n",
    "        msprime.PopulationParametersChange(\n",
    "            population_id=1, time=10, growth_rate=r_EU\n",
    "        ),\n",
    "        msprime.PopulationParametersChange(\n",
    "            population_id=2, time=10, growth_rate=r_AS\n",
    "        ),\n",
    "        \n",
    "    ]\n",
    "    \n",
    "    # admixture\n",
    "    admixture = [\n",
    "        msprime.MassMigration(time=12, source=4, destination=0, proportion=.15),\n",
    "        msprime.MassMigration(time=12, source=4, destination=1, proportion=.15),\n",
    "        msprime.MassMigration(time=12, source=4, destination=2, proportion=.30),\n",
    "        msprime.MassMigration(time=12, source=4, destination=3, proportion=.30),\n",
    "    ]\n",
    "    \n",
    "\n",
    "    # JPT splits from CHB\n",
    "    JPT_event = [\n",
    "        # seg migration rate JPT <--> CHB to zero\n",
    "        msprime.MigrationRateChange(time=T_CH_JP, rate=0.0, matrix_index=(2, 3)),\n",
    "        msprime.MigrationRateChange(time=T_CH_JP, rate=0.0, matrix_index=(3, 2)),\n",
    "        # JPT splits from CHB\n",
    "        msprime.MassMigration(time=T_CH_JP, source=3, destination=2, proportion=1.0),\n",
    "    ]\n",
    "\n",
    "    # CEU and CHB merge into B\n",
    "    AS_event = [\n",
    "        # set migration rates to zero\n",
    "        msprime.MigrationRateChange(time=T_EU_AS, rate=0.0),\n",
    "        # CHB splits from CEU\n",
    "        msprime.MassMigration(time=T_EU_AS, source=2, destination=1, proportion=1.0),\n",
    "        # change size of B to N_B\n",
    "        # set growth rates in B to zero\n",
    "        msprime.PopulationParametersChange(\n",
    "            time=T_EU_AS, initial_size=N_B, growth_rate=0.0, population_id=1\n",
    "        ),\n",
    "        # set up migration between YRI and B\n",
    "        msprime.MigrationRateChange(time=T_EU_AS, rate=m_AF_B, matrix_index=(0, 1)),\n",
    "        msprime.MigrationRateChange(time=T_EU_AS, rate=m_AF_B, matrix_index=(1, 0)),\n",
    "    ]\n",
    "\n",
    "    # Out-of-Africa event\n",
    "    EU_event = [\n",
    "        # B splits from YRI\n",
    "        msprime.MigrationRateChange(time=T_B, rate=0.0),\n",
    "        msprime.MassMigration(time=T_B, source=1, destination=0, proportion=1.0),\n",
    "    ]\n",
    "\n",
    "    # Change size of YRI/African population\n",
    "    AF_event = [\n",
    "        msprime.PopulationParametersChange(time=T_AF, initial_size=N_A, population_id=0)\n",
    "    ]\n",
    "\n",
    "    demographic_events = stop_recent_growth + admixture + JPT_event + AS_event + EU_event + AF_event\n",
    "\n",
    "    return stdpopsim.DemographicModel(\n",
    "        id=id,\n",
    "        description=description,\n",
    "        long_description=long_description,\n",
    "        populations=populations,\n",
    "        citations=citations,\n",
    "        generation_time=generation_time,\n",
    "        #mutation_rate=mutation_rate,\n",
    "        population_configurations=population_configurations,\n",
    "        migration_matrix=migration_matrix,\n",
    "        demographic_events=demographic_events,\n",
    "    )\n",
    "\n",
    "\n",
    "_species.add_demographic_model(_ooa_4popADMIXTURE())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c8570ad1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DemographyDebugger\n",
      "╠═══════════════════════════════╗\n",
      "║ Epoch[0]: [0, 10) generations ║\n",
      "╠═══════════════════════════════╝\n",
      "╟    Populations (total=5 active=5)\n",
      "║    ┌────────────────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │      start│        end│growth_rate  │  pop_0   │  pop_1   │  pop_2   │  pop_3  │ pop_4 │\n",
      "║    ├────────────────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│   100000.0│    23721.0│ 0.144       │    0     │ 1.14e-05 │ 5.6e-06  │    0    │   0   │\n",
      "║    │  pop_1│   100000.0│    23721.0│ 0.144       │ 1.14e-05 │    0     │ 4.75e-05 │    0    │   0   │\n",
      "║    │  pop_2│   100000.0│    62955.1│ 0.0463      │ 5.6e-06  │ 4.75e-05 │    0     │ 3.3e-05 │   0   │\n",
      "║    │  pop_3│   239119.1│   210179.4│ 0.0129      │    0     │    0     │ 3.3e-05  │    0    │   0   │\n",
      "║    │  pop_4│    91105.9│    55258.5│ 0.05        │    0     │    0     │    0     │    0    │   0   │\n",
      "║    └────────────────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 10\n",
      "║    ┌────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │  time│type        │parameters          │effect                                 │\n",
      "║    ├────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │    10│Population  │population=0,       │growth_rate → 0 for population 0       │\n",
      "║    │      │parameter   │growth_rate=0       │                                       │\n",
      "║    │      │change      │                    │                                       │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │    10│Population  │population=1,       │growth_rate → 0.0016 for population 1  │\n",
      "║    │      │parameter   │growth_rate=0.0016  │                                       │\n",
      "║    │      │change      │                    │                                       │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │    10│Population  │population=2,       │growth_rate → 0.0026 for population 2  │\n",
      "║    │      │parameter   │growth_rate=0.0026  │                                       │\n",
      "║    │      │change      │                    │                                       │\n",
      "║    └────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠════════════════════════════════╗\n",
      "║ Epoch[1]: [10, 12) generations ║\n",
      "╠════════════════════════════════╝\n",
      "╟    Populations (total=5 active=5)\n",
      "║    ┌────────────────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │      start│        end│growth_rate  │  pop_0   │  pop_1   │  pop_2   │  pop_3  │ pop_4 │\n",
      "║    ├────────────────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│    23721.0│    23721.0│ 0           │    0     │ 1.14e-05 │ 5.6e-06  │    0    │   0   │\n",
      "║    │  pop_1│    23721.0│    23645.2│ 0.0016      │ 1.14e-05 │    0     │ 4.75e-05 │    0    │   0   │\n",
      "║    │  pop_2│    62955.1│    62628.6│ 0.0026      │ 5.6e-06  │ 4.75e-05 │    0     │ 3.3e-05 │   0   │\n",
      "║    │  pop_3│   210179.4│   204826.1│ 0.0129      │    0     │    0     │ 3.3e-05  │    0    │   0   │\n",
      "║    │  pop_4│    55258.5│    50000.0│ 0.05        │    0     │    0     │    0     │    0    │   0   │\n",
      "║    └────────────────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 12\n",
      "║    ┌──────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │  time│type            │parameters         │effect                                │\n",
      "║    ├──────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │    12│Mass Migration  │source=4, dest=0,  │Lineages currently in population 4    │\n",
      "║    │      │                │proportion=0.15    │move to 0 with probability 0.15       │\n",
      "║    │      │                │                   │(equivalent to individuals migrating  │\n",
      "║    │      │                │                   │from 0 to 4 forwards in time)         │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │    12│Mass Migration  │source=4, dest=1,  │Lineages currently in population 4    │\n",
      "║    │      │                │proportion=0.15    │move to 1 with probability 0.15       │\n",
      "║    │      │                │                   │(equivalent to individuals migrating  │\n",
      "║    │      │                │                   │from 1 to 4 forwards in time)         │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │    12│Mass Migration  │source=4, dest=2,  │Lineages currently in population 4    │\n",
      "║    │      │                │proportion=0.3     │move to 2 with probability 0.3        │\n",
      "║    │      │                │                   │(equivalent to individuals migrating  │\n",
      "║    │      │                │                   │from 2 to 4 forwards in time)         │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │    12│Mass Migration  │source=4, dest=3,  │Lineages currently in population 4    │\n",
      "║    │      │                │proportion=0.3     │move to 3 with probability 0.3        │\n",
      "║    │      │                │                   │(equivalent to individuals migrating  │\n",
      "║    │      │                │                   │from 3 to 4 forwards in time)         │\n",
      "║    └──────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠═════════════════════════════════╗\n",
      "║ Epoch[2]: [12, 310) generations ║\n",
      "╠═════════════════════════════════╝\n",
      "╟    Populations (total=5 active=5)\n",
      "║    ┌───────────────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │      start│       end│growth_rate  │  pop_0   │  pop_1   │  pop_2   │  pop_3  │ pop_4 │\n",
      "║    ├───────────────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│    23721.0│   23721.0│ 0           │    0     │ 1.14e-05 │ 5.6e-06  │    0    │   0   │\n",
      "║    │  pop_1│    23645.2│   14678.2│ 0.0016      │ 1.14e-05 │    0     │ 4.75e-05 │    0    │   0   │\n",
      "║    │  pop_2│    62628.6│   28859.0│ 0.0026      │ 5.6e-06  │ 4.75e-05 │    0     │ 3.3e-05 │   0   │\n",
      "║    │  pop_3│   204826.1│    4384.0│ 0.0129      │    0     │    0     │ 3.3e-05  │    0    │   0   │\n",
      "║    │  pop_4│    50000.0│       0.0│ 0.05        │    0     │    0     │    0     │    0    │   0   │\n",
      "║    └───────────────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 310\n",
      "║    ┌────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │  time│type            │parameters         │effect                                  │\n",
      "║    ├────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │   310│Migration rate  │source=2, dest=3,  │Backwards-time migration rate from 2    │\n",
      "║    │      │change          │rate=0.0           │to 3 → 0.0                              │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │   310│Migration rate  │source=3, dest=2,  │Backwards-time migration rate from 3    │\n",
      "║    │      │change          │rate=0.0           │to 2 → 0.0                              │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │   310│Mass Migration  │source=3, dest=2,  │All lineages currently in population 3  │\n",
      "║    │      │                │proportion=1       │move to 2 (equivalent to individuals    │\n",
      "║    │      │                │                   │migrating from 2 to 3 forwards in       │\n",
      "║    │      │                │                   │time)                                   │\n",
      "║    └────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠═══════════════════════════════════════╗\n",
      "║ Epoch[3]: [310, 1.59e+03) generations ║\n",
      "╠═══════════════════════════════════════╝\n",
      "╟    Populations (total=5 active=5)\n",
      "║    ┌────────────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │     start│       end│growth_rate  │  pop_0   │  pop_1   │  pop_2   │ pop_3 │ pop_4 │\n",
      "║    ├────────────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│   23721.0│   23721.0│ 0           │    0     │ 1.14e-05 │ 5.6e-06  │   0   │   0   │\n",
      "║    │  pop_1│   14678.2│    1905.5│ 0.0016      │ 1.14e-05 │    0     │ 4.75e-05 │   0   │   0   │\n",
      "║    │  pop_2│   28859.0│    1045.8│ 0.0026      │ 5.6e-06  │ 4.75e-05 │    0     │   0   │   0   │\n",
      "║    │  pop_3│    4384.0│       0.0│ 0.0129      │    0     │    0     │    0     │   0   │   0   │\n",
      "║    │  pop_4│       0.0│       0.0│ 0.05        │    0     │    0     │    0     │   0   │   0   │\n",
      "║    └────────────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 1.59e+03\n",
      "║    ┌──────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │  time│type            │parameters           │effect                                  │\n",
      "║    ├──────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  1586│Migration rate  │source=-1, dest=-1,  │Backwards-time migration rate for all   │\n",
      "║    │      │change          │rate=0.0             │populations → 0.0                       │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  1586│Mass Migration  │source=2, dest=1,    │All lineages currently in population 2  │\n",
      "║    │      │                │proportion=1         │move to 1 (equivalent to individuals    │\n",
      "║    │      │                │                     │migrating from 1 to 2 forwards in       │\n",
      "║    │      │                │                     │time)                                   │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  1586│Population      │population=1,        │initial_size → 2.8e+03 and growth_rate  │\n",
      "║    │      │parameter       │initial_size=2831,   │→ 0 for population 1                    │\n",
      "║    │      │change          │growth_rate=0.0      │                                        │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  1586│Migration rate  │source=0, dest=1,    │Backwards-time migration rate from 0    │\n",
      "║    │      │change          │rate=0.000168        │to 1 → 0.000168                         │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  1586│Migration rate  │source=1, dest=0,    │Backwards-time migration rate from 1    │\n",
      "║    │      │change          │rate=0.000168        │to 0 → 0.000168                         │\n",
      "║    └──────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠═══════════════════════════════════════════╗\n",
      "║ Epoch[4]: [1.59e+03, 4.1e+03) generations ║\n",
      "╠═══════════════════════════════════════════╝\n",
      "╟    Populations (total=5 active=5)\n",
      "║    ┌─────────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │     start│       end│growth_rate  │  pop_0   │  pop_1   │ pop_2 │ pop_3 │ pop_4 │\n",
      "║    ├─────────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│   23721.0│   23721.0│ 0           │    0     │ 0.000168 │   0   │   0   │   0   │\n",
      "║    │  pop_1│    2831.0│    2831.0│ 0           │ 0.000168 │    0     │   0   │   0   │   0   │\n",
      "║    │  pop_2│    1045.8│       1.5│ 0.0026      │    0     │    0     │   0   │   0   │   0   │\n",
      "║    │  pop_3│       0.0│       0.0│ 0.0129      │    0     │    0     │   0   │   0   │   0   │\n",
      "║    │  pop_4│       0.0│       0.0│ 0.05        │    0     │    0     │   0   │   0   │   0   │\n",
      "║    └─────────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 4.1e+03\n",
      "║    ┌──────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │  time│type            │parameters           │effect                                  │\n",
      "║    ├──────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  4103│Migration rate  │source=-1, dest=-1,  │Backwards-time migration rate for all   │\n",
      "║    │      │change          │rate=0.0             │populations → 0.0                       │\n",
      "║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│\n",
      "║    │  4103│Mass Migration  │source=1, dest=0,    │All lineages currently in population 1  │\n",
      "║    │      │                │proportion=1         │move to 0 (equivalent to individuals    │\n",
      "║    │      │                │                     │migrating from 0 to 1 forwards in       │\n",
      "║    │      │                │                     │time)                                   │\n",
      "║    └──────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠═══════════════════════════════════════════╗\n",
      "║ Epoch[5]: [4.1e+03, 1.23e+04) generations ║\n",
      "╠═══════════════════════════════════════════╝\n",
      "╟    Populations (total=5 active=5)\n",
      "║    ┌───────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │     start│       end│growth_rate  │ pop_0 │ pop_1 │ pop_2 │ pop_3 │ pop_4 │\n",
      "║    ├───────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│   23721.0│   23721.0│ 0           │   0   │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_1│    2831.0│    2831.0│ 0           │   0   │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_2│       1.5│       0.0│ 0.0026      │   0   │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_3│       0.0│       0.0│ 0.0129      │   0   │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_4│       0.0│       0.0│ 0.05        │   0   │   0   │   0   │   0   │   0   │\n",
      "║    └───────────────────────────────────────────────────────────────────────────────────┘\n",
      "╟    Events @ generation 1.23e+04\n",
      "║    ┌─────────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       time│type        │parameters          │effect                                 │\n",
      "║    ├─────────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  1.231e+04│Population  │population=0,       │initial_size → 1.1e+04 for population  │\n",
      "║    │           │parameter   │initial_size=11293  │0                                      │\n",
      "║    │           │change      │                    │                                       │\n",
      "║    └─────────────────────────────────────────────────────────────────────────────────────┘\n",
      "╠═══════════════════════════════════════╗\n",
      "║ Epoch[6]: [1.23e+04, inf) generations ║\n",
      "╠═══════════════════════════════════════╝\n",
      "╟    Populations (total=5 active=5)\n",
      "║    ┌───────────────────────────────────────────────────────────────────────────────────┐\n",
      "║    │       │     start│       end│growth_rate  │ pop_0 │ pop_1 │ pop_2 │ pop_3 │ pop_4 │\n",
      "║    ├───────────────────────────────────────────────────────────────────────────────────┤\n",
      "║    │  pop_0│   11293.0│   11293.0│ 0           │   0   │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_1│    2831.0│    2831.0│ 0           │   0   │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_2│       0.0│       0.0│ 0.0026      │   0   │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_3│       0.0│       0.0│ 0.0129      │   0   │   0   │   0   │   0   │   0   │\n",
      "║    │  pop_4│       0.0│       0.0│ 0.05        │   0   │   0   │   0   │   0   │   0   │\n",
      "║    └───────────────────────────────────────────────────────────────────────────────────┘\n"
     ]
    }
   ],
   "source": [
    "_species.get_demographic_model(\"OutOfAfrica_4J17_ADMIXTURE\").debug()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d5053b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de8768e3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8692f149",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name = \"HomSap\"\n",
    "model_name = \"OutOfAfrica_4J17_ADMIXTURE\"\n",
    "chromosome = \"chr22\"\n",
    "\n",
    "remember_gen = 20\n",
    "ancestral_Ne = 11293\n",
    "mut_rate = 1.44e-8\n",
    "\n",
    "length_multiplier=.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "83629db4",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = stdpopsim.get_species(species_name)\n",
    "model = species.get_demographic_model(model_name)\n",
    "contig = species.get_contig(chromosome, length_multiplier=length_multiplier)  # default is a flat genetic map\n",
    "newmap = msprime.RecombinationMap(\n",
    "    positions=contig.recombination_map.get_positions(),\n",
    "    rates= [x*3 for x in contig.recombination_map.get_rates()]\n",
    ")\n",
    "\n",
    "newcontig = stdpopsim.Contig(recombination_map = newmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1a05c238",
   "metadata": {},
   "outputs": [],
   "source": [
    "WORK_DIR = '/home/kele/Documents/lai/test/test_4pop'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8f7b4bfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/kele/Documents/lai/test/test_4pop/run.slim /home/kele/Documents/lai/test/test_4pop/from_slim.ts\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/home/kele/Documents/lai/test/test_4pop’: File exists\n"
     ]
    }
   ],
   "source": [
    "os.system(f'mkdir {WORK_DIR}')\n",
    "SLiM_script_path = os.path.join(WORK_DIR, 'run.slim')\n",
    "SLiM_ts_path = os.path.join(WORK_DIR, 'from_slim.ts')\n",
    "print(SLiM_script_path, SLiM_ts_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "590844bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100000\n",
      "100000\n",
      "100000\n",
      "239119.05109482087\n",
      "91105.94001952546\n"
     ]
    }
   ],
   "source": [
    "for i in range(model.num_populations):\n",
    "    print(model.population_configurations[i].initial_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f8cb7e71",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = model.get_samples(100000,100000,100000,20000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d4ff978a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    initializeGeneConversion(0.6666666, 300, 1.0);\n",
      "    sim.registerLateEvent(\"s000000\", \"{dbg(self.source); sim.treeSeqRememberIndividuals(sim.subpopulations.individuals);}\", G_end-20, G_end-20);\n"
     ]
    }
   ],
   "source": [
    "engine = stdpopsim.get_engine(\"slim\")\n",
    "with open(SLiM_script_path, \"w\") as f:\n",
    "    with redirect_stdout(f):\n",
    "        _ = engine.simulate(\n",
    "            model,\n",
    "            newcontig,\n",
    "            samples=samples,\n",
    "            slim_script=True,\n",
    "            verbosity=2\n",
    "        )\n",
    "\n",
    "# add gene conversion\n",
    "os.system(f\"sed -i 's+(recombination_rates, recombination_ends);+(recombination_rates, recombination_ends);\\\\n    initializeGeneConversion(0.6666666, 300, 1.0);+g' {SLiM_script_path}\")\n",
    "\n",
    "\n",
    "# check the edits\n",
    "os.system(f'grep {SLiM_ts_path} {SLiM_script_path}')\n",
    "os.system(f\"grep 'initializeGeneConversion' {SLiM_script_path}\")\n",
    "\n",
    "\n",
    "# The goal is to mark all inds 20 generations prior to the present.\n",
    "# Then use these marked individuals to define the \"local ancestry\".\n",
    "# G_end is a variable define in the SLiM script that\n",
    "# should give the end of the simulation (in generations)\n",
    "\n",
    "\n",
    "remember_cmd = r's+// Admixture pulses\\.+sim.registerLateEvent(\"s000000\", \"{dbg(self.source); sim.treeSeqRememberIndividuals(sim.subpopulations.individuals);}\", G_end-20, G_end-20);\\n    // Admixture pulses\\.+g '\n",
    "remember_cmd\n",
    "\n",
    "\n",
    "os.system(f\"sed -i '{remember_cmd}' {SLiM_script_path}\")\n",
    "\n",
    "\n",
    "# check the edit\n",
    "os.system(f\"grep 's000000' {SLiM_script_path}\")\n",
    "\n",
    "import fileinput\n",
    "for line in fileinput.input(f'{SLiM_script_path}', inplace=True):\n",
    "    line = line.rstrip('\\r\\n')\n",
    "    if line.strip().startswith('''defineConstant(\"trees_file\",'''):\n",
    "        print(f'    //defineConstant(\"trees_file\", \"{os.path.join(WORK_DIR, \"from_slim.ts\")});')\n",
    "    else:\n",
    "        print(line)\n",
    "fileinput.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcc0b527",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3809e422",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f31caa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters\n",
    "\n",
    "run_slim = False\n",
    "run_la = False\n",
    "\n",
    "\n",
    "npop = 4\n",
    "admixed_pop_idx = 3\n",
    "nref = 150000\n",
    "\n",
    "SEED = 321\n",
    "\n",
    "WORK_DIR = f'/home/users/waplesr/lai/large/AmericanAdmixture_4B11_100K/{SEED}'\n",
    "BCFTOOLS = '/home/users/waplesr/lai/large/programs/hts/bcftools-1.13/bcftools'\n",
    "\n",
    "\n",
    "# ## Species and demographic model\n",
    "\n",
    "species = stdpopsim.get_species(demo_species)\n",
    "model = species.get_demographic_model(demo_model)\n",
    "\n",
    "\n",
    "# ## Genome specification\n",
    "# We will use **contig** when simulating in SLiM, this simulation is with gene conversion,\n",
    "# **newcontig** is for use in msprime, and will not include gene conversion.\n",
    "\n",
    "contig = species.get_contig(chromosome, length_multiplier=length_multiplier)  # default is a flat genetic map\n",
    "\n",
    "# make a new contig\n",
    "# constant recombination rate, but 3 times higher rec rate than normal\n",
    "# this is to allow 2/3rds of rec events to be gene conversions.\n",
    "newmap = msprime.RecombinationMap(\n",
    "    positions=contig.recombination_map.get_positions(),\n",
    "    rates= [x*3 for x in contig.recombination_map.get_rates()]\n",
    ")\n",
    "\n",
    "newcontig = stdpopsim.Contig(recombination_map = newmap)\n",
    "\n",
    "os.system(f'mkdir {WORK_DIR}')\n",
    "SLiM_script_path = os.path.join(WORK_DIR, 'run.slim')\n",
    "SLiM_ts_path = os.path.join(WORK_DIR, 'from_slim.ts')\n",
    "print(SLiM_script_path, SLiM_ts_path)\n",
    "\n",
    "\n",
    "model.population_configurations[0].initial_size,model.population_configurations[1].initial_size,model.population_configurations[2].initial_size\n",
    "\n",
    "def get_growth_rate(initial_size, final_size, generations):\n",
    "    return np.log(final_size/initial_size)/generations\n",
    "\n",
    "pop0_recent_growth = get_growth_rate(model.population_configurations[0].initial_size, 100000, 10)\n",
    "pop1_recent_growth = get_growth_rate(model.population_configurations[1].initial_size, 100000, 10)\n",
    "pop2_recent_growth = get_growth_rate(model.population_configurations[2].initial_size, 100000, 10)\n",
    "\n",
    "model.population_configurations[0].initial_size = 100000\n",
    "model.population_configurations[1].initial_size = 100000\n",
    "model.population_configurations[2].initial_size = 100000\n",
    "\n",
    "recent_growth = [\n",
    " msprime.PopulationParametersChange(population=0, time=0, growth_rate=pop0_recent_growth),\n",
    " msprime.PopulationParametersChange(population=1, time=0, growth_rate=pop1_recent_growth),\n",
    " msprime.PopulationParametersChange(population=2, time=0, growth_rate=pop2_recent_growth),\n",
    "]\n",
    "\n",
    "recent_growth = recent_growth + [\n",
    " msprime.PopulationParametersChange(population=0, time=10, growth_rate=0),\n",
    " msprime.PopulationParametersChange(population=1, time=10, growth_rate=0.0038),\n",
    " msprime.PopulationParametersChange(population=2, time=10, growth_rate=0.0048),\n",
    "]\n",
    "\n",
    "model.demographic_events = recent_growth + model.demographic_events\n",
    "\n",
    "\n",
    "samples = model.get_samples(100000,100000,100000,20000)\n",
    "\n",
    "\n",
    "# ## Samples\n",
    "\n",
    "remake_script = False\n",
    "\n",
    "if remake_script:\n",
    "    engine = stdpopsim.get_engine(\"slim\")\n",
    "    with open(SLiM_script_path, \"w\") as f:\n",
    "        with redirect_stdout(f):\n",
    "            _ = engine.simulate(\n",
    "                model,\n",
    "                newcontig,\n",
    "                samples=samples,\n",
    "                seed=SEED, # not sure if the seed is recognised here\n",
    "                slim_script=True,\n",
    "                verbosity=2\n",
    "            )\n",
    "\n",
    "\n",
    "            ## Editing the SLiM script prior to execution\n",
    "\n",
    "            # specify the output file and set up gene conversion\n",
    "\n",
    "            # change the ts output location\n",
    "            import fileinput\n",
    "            for line in fileinput.input(f'{SLiM_script_path}', inplace=True):\n",
    "                line = line.rstrip('\\r\\n')\n",
    "                if line.startswith('''defineConstant(\"trees_file\",'''):\n",
    "                    print(f'defineConstant(\"trees_file\", \"{os.path.join(WORK_DIR, \"from_slim.ts\")});')\n",
    "                else:\n",
    "                    print(line)\n",
    "            fileinput.close()\n",
    "\n",
    "            #sed_cmd = f's+\"/scratch/*.*.lindstroem.q/tmp*.ts+\"{SLiM_ts_path}+g'\n",
    "            #os.system(f'sed -i {sed_cmd} {SLiM_script_path}')\n",
    "\n",
    "            # os.system(f\"sed -i 's@*scratch*@defineConstant(\"trees_file\", \"/home/users/waplesr/lai/large/AmericanAdmixture_4B11_100K/1234567890/from_slim.ts\")@g'   {SLiM_script_path}\")\n",
    "\n",
    "            # add gene conversion\n",
    "            os.system(f\"sed -i 's+(recombination_rates, recombination_ends);+(recombination_rates, recombination_ends);\\\\n    initializeGeneConversion(0.6666666, 300, 1.0);+g' {SLiM_script_path}\")\n",
    "\n",
    "\n",
    "            # check the edits\n",
    "            os.system(f'grep {SLiM_ts_path} {SLiM_script_path}')\n",
    "            os.system(f\"grep 'initializeGeneConversion' {SLiM_script_path}\")\n",
    "\n",
    "\n",
    "            # The goal is to mark all inds 20 generations prior to the present.\n",
    "            # Then use these marked individuals to define the \"local ancestry\".\n",
    "            # G_end is a variable define in the SLiM script that\n",
    "            # should give the end of the simulation (in generations)\n",
    "\n",
    "\n",
    "            remember_cmd = r's+// Admixture pulses\\.+sim.registerLateEvent(\"s000000\", \"{dbg(self.source); sim.treeSeqRememberIndividuals(sim.subpopulations.individuals);}\", G_end-20, G_end-20);\\n    // Admixture pulses\\.+g '\n",
    "            remember_cmd\n",
    "\n",
    "\n",
    "            os.system(f\"sed -i '{remember_cmd}' {SLiM_script_path}\")\n",
    "\n",
    "\n",
    "            # check the edit\n",
    "            os.system(f\"grep 's000000' {SLiM_script_path}\")\n",
    "\n",
    "\n",
    "if run_slim:\n",
    "    os.system(f'/home/users/waplesr/lai/large/programs/slim/build/slim -seed {SEED} {SLiM_script_path} > {SLiM_script_path}.log')\n",
    "\n",
    "# ## Load in the ts from SLiM\n",
    "\n",
    "\n",
    "ts = tskit.load(SLiM_ts_path)\n",
    "ts\n",
    "\n",
    "\n",
    "# ## Recapitate\n",
    "\n",
    "# #### Use the original recombination rates for the recapitation, rather than the elevated rates use for gene conversion\n",
    "# #### Notice this is also a discrete genetic map.\n",
    "\n",
    "recapmap = msprime.RecombinationMap(\n",
    "    positions=[0.0, ts.get_sequence_length()],\n",
    "    rates= contig.recombination_map.get_rates(),\n",
    "    num_loci = int(ts.get_sequence_length())\n",
    ")\n",
    "\n",
    "\n",
    "# 7310 is the ancestral population size in Africa. All the action should be occuring in Africa here.\n",
    "#\n",
    "# Recapitation uses the Hudson coalescent.\n",
    "\n",
    "coalesced_ts = msprime.simulate(\n",
    "    Ne=ancestral_Ne,\n",
    "    from_ts=ts,\n",
    "    random_seed=SEED,\n",
    "    recombination_map=recapmap,\n",
    "    population_configurations = [\n",
    "        msprime.PopulationConfiguration(),\n",
    "        msprime.PopulationConfiguration(),\n",
    "        msprime.PopulationConfiguration(),\n",
    "        msprime.PopulationConfiguration()\n",
    "    ],\n",
    "    migration_matrix=None\n",
    "    )\n",
    "\n",
    "del ts\n",
    "\n",
    "\n",
    "for x in range(4):\n",
    "    y = coalesced_ts.tables.nodes.time[np.where(coalesced_ts.tables.nodes.population==x)[0]]\n",
    "    plt.plot(np.log10(np.sort(y)+1), label = x)\n",
    "plt.legend(title = 'population')\n",
    "plt.ylabel('log10(node age+1)')\n",
    "plt.xlabel('nodes within each population')\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# ## Here I need to edit the ts so that only nodes sampled in the present are marked as 'samples'\n",
    "\n",
    "tables = coalesced_ts.tables\n",
    "newnodes = tables.nodes.copy()\n",
    "\n",
    "flags = tables.nodes.flags\n",
    "assert flags.sum() == coalesced_ts.num_samples\n",
    "\n",
    "flags[tables.nodes.time == remember_gen] = 0\n",
    "assert flags.sum() == np.intersect1d(\n",
    "                    coalesced_ts.samples(),\n",
    "                    np.where(tables.nodes.asdict()['time']==0)[0]\n",
    "                ).size\n",
    "\n",
    "newnodes.set_columns(\n",
    "    time = newnodes.time,\n",
    "    flags=flags,\n",
    "    population=newnodes.population,\n",
    "    individual=newnodes.individual,\n",
    "    metadata=newnodes.metadata,\n",
    "    metadata_offset=newnodes.metadata_offset,\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "tables.nodes.clear()\n",
    "\n",
    "\n",
    "for row in newnodes:\n",
    "    tables.nodes.add_row(flags=row.flags,\n",
    "                         time=row.time,\n",
    "                         population=row.population,\n",
    "                         individual=row.individual,\n",
    "                         metadata=row.metadata)\n",
    "\n",
    "\n",
    "coalesced_ts = tables.tree_sequence()\n",
    "del newnodes\n",
    "del tables\n",
    "del flags\n",
    "\n",
    "for x in range(4):\n",
    "    print (int(len(coalesced_ts.samples(x))/2))\n",
    "\n",
    "\n",
    "# downsample the coalesced ts\n",
    "\n",
    "\n",
    "for pop, nsamp in enumerate([100000,100000,100000,20000]):\n",
    "    print(pop, nsamp)\n",
    "    coalesced_ts.samples(pop)[:nsamp]\n",
    "\n",
    "\n",
    "\n",
    "# ## Add mutations\n",
    "# Also remove sites with a minor allele count (MAC) of 1. Notice this is removing sites that have MAC=1 at the population level, so a subset of these individuals may still have singletons.\n",
    "\n",
    "\n",
    "\n",
    "mut_ts = msprime.mutate(\n",
    "    tree_sequence=coalesced_ts,\n",
    "    rate=mut_rate,\n",
    "    random_seed=SEED,\n",
    "    model = msprime.InfiniteSites(alphabet=msprime.NUCLEOTIDES))\n",
    "mut_ts\n",
    "\n",
    "\n",
    "tszip.compress(mut_ts, os.path.join(WORK_DIR, 'mut_ts.tsz'))\n",
    "\n",
    "mut_ts.dump(os.path.join(WORK_DIR, 'mut_ts.ts'))\n",
    "\n",
    "\n",
    "# # Filter sites\n",
    "\n",
    "# # Validation and model checking\n",
    "\n",
    "\n",
    "site_fst_of_pops = dict()\n",
    "for x in range(4):\n",
    "    for y in range(x,4):\n",
    "        site_fst_of_pops[(x,y)] = mut_ts.Fst(sample_sets = ([mut_ts.samples(population=x), mut_ts.samples(population=y)]))\n",
    "\n",
    "print(site_fst_of_pops)\n",
    "\n",
    "for pop in range(4):\n",
    "    print (int((len(mut_ts.samples(population=pop))/2)))\n",
    "\n",
    "\n",
    "# # Filtering sites\n",
    "\n",
    "mut_ts = tszip.decompress(os.path.join(WORK_DIR, 'mut_ts.tsz'))\n",
    "mac1_ts = strip_singletons(mut_ts)\n",
    "mac1_ts = strip_adjacent_sites(mac1_ts)\n",
    "del mut_ts\n",
    "\n",
    "\n",
    "# # get local ancestry tracts\n",
    "# needs to work on the entire ts\n",
    "\n",
    "if run_la:\n",
    "    #local_df = get_local_ancestry_pop_multi(\n",
    "    local_df = get_local_ancestry_pop(\n",
    "        ts=mac1_ts,\n",
    "        pop=3,\n",
    "        admixture_time=remember_gen,\n",
    "        max_samples=20000,\n",
    "        per_rep=12,\n",
    "        #n_cores=10\n",
    "    )\n",
    "\n",
    "    local_df.to_hdf(os.path.join(WORK_DIR, 'la_df.h5'), key='la_df', complib='blosc:lz4', complevel=9)\n",
    "\n",
    "\n",
    "\n",
    "local_df = pd.read_hdf(os.path.join(WORK_DIR, 'la_df.h5'), key='la_df')\n",
    "simp_ts, mapping = mac1_ts.simplify(\n",
    "    samples=local_df['child'].unique().tolist(),\n",
    "    map_nodes=True,\n",
    "    filter_populations=False\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "# remove singletons from the target samples\n",
    "simp_ts = strip_singletons(simp_ts)\n",
    "\n",
    "\n",
    "if run_la:\n",
    "    la_mat, samps = get_la_mat_large(\n",
    "        ts=simp_ts,\n",
    "        df=local_df,\n",
    "        mapping=mapping\n",
    "    )\n",
    "    np.savez_compressed(os.path.join(WORK_DIR, 'la_mat.npz'), la_mat=la_mat)\n",
    "\n",
    "la_mat = np.load(os.path.join(WORK_DIR, 'la_mat.npz'))['la_mat']\n",
    "plt.plot(la_mat[:,1])\n",
    "\n",
    "\n",
    "plt.subplots(figsize = (16,16))\n",
    "sns.heatmap(la_mat[::100].T)\n",
    "\n",
    "\n",
    "# # Export the data to vcf\n",
    "\n",
    "\n",
    "keep = np.concatenate(\n",
    "    [mac1_ts.samples(pop)[:nsamp] for pop, nsamp in enumerate([100000,100000,100000,20000])]\n",
    ")\n",
    "\n",
    "downsampled_ts = mac1_ts.simplify(keep)\n",
    "downsampled_ts\n",
    "\n",
    "\n",
    "keep_positions = set([s.position for s in simp_ts.sites()])\n",
    "\n",
    "remove_positions = [i for (i,s) in enumerate(downsampled_ts.sites()) if s.position not in keep_positions]\n",
    "\n",
    "export_ts = downsampled_ts.delete_sites(remove_positions)\n",
    "assert export_ts.num_sites == la_mat.shape[0]\n",
    "\n",
    "\n",
    "# update the individuals table\n",
    "tables = export_ts.tables\n",
    "tables.individuals.clear()\n",
    "tables.nodes.individual = np.repeat(np.array([-1], dtype = 'int32'), len(tables.nodes))\n",
    "export_ts = tables.tree_sequence()\n",
    "\n",
    "\n",
    "pop_of_sample = dict(zip(range(len(export_ts.tables.nodes)), export_ts.tables.nodes.population))\n",
    "pop_of_sample\n",
    "\n",
    "nind_in_pop = collections.defaultdict(int)\n",
    "pops = [pop_of_sample[i] for i in export_ts.samples()]\n",
    "for p in pops:\n",
    "    nind_in_pop[p] +=1\n",
    "for p in nind_in_pop:\n",
    "    nind_in_pop[p] = int(nind_in_pop[p]/2)\n",
    "ind_labels = []\n",
    "for p in nind_in_pop:\n",
    "    for i in range(1, nind_in_pop[p]+1):\n",
    "        ind_labels.append(f'pop_{p}-ind_{i:05}')\n",
    "len(ind_labels)\n",
    "\n",
    "\n",
    "# # Write bcf file\n",
    "\n",
    "tszip.compress(export_ts, os.path.join(WORK_DIR, 'export_ts.tsz'))\n",
    "\n",
    "export_ts = tszip.decompress(os.path.join(WORK_DIR, 'export_ts.tsz'))\n",
    "\n",
    "read_fd, write_fd = os.pipe()\n",
    "write_pipe = os.fdopen(write_fd, \"w\")\n",
    "with open(os.path.join(WORK_DIR, 'genotypes.bcf'), \"w\") as bcf_file:\n",
    "    proc = subprocess.Popen(\n",
    "        [f\"{BCFTOOLS}\", \"view\", \"-O\", \"b\"], stdin=read_fd, stdout=bcf_file\n",
    "    )\n",
    "    export_ts.write_vcf(write_pipe, ploidy=2, contig_id='22', individual_names=ind_labels)\n",
    "    write_pipe.close()\n",
    "    os.close(read_fd)\n",
    "    proc.wait()\n",
    "    if proc.returncode != 0:\n",
    "        raise RuntimeError(\"bcftools failed with status:\", proc.returncode)\n",
    "\n",
    "\n",
    "with open(os.path.join(WORK_DIR, 'reference_inds.txt'), 'w') as OUTFILE:\n",
    "    for ind in ind_labels[:nref]:\n",
    "        OUTFILE.write(f'{ind}\\n')\n",
    "\n",
    "with open(os.path.join(WORK_DIR, 'target_inds.txt'), 'w') as OUTFILE:\n",
    "    for ind in ind_labels[nref:]:\n",
    "        OUTFILE.write(f'{ind}\\n')\n",
    "\n",
    "os.system(f\"{BCFTOOLS} view --samples-file {os.path.join(WORK_DIR, 'target_inds.txt')} --output-type z --output {os.path.join(WORK_DIR, 'target_inds.vcf.gz')} {os.path.join(WORK_DIR, 'genotypes.bcf')}\")\n",
    "os.system(f\"{BCFTOOLS} view --samples-file {os.path.join(WORK_DIR, 'reference_inds.txt')} --output-type z --output {os.path.join(WORK_DIR, 'reference_inds.vcf.gz')} {os.path.join(WORK_DIR, 'genotypes.bcf')}\")\n",
    "\n",
    "\n",
    "# # Make a genetic map file\n",
    "\n",
    "os.system(f\"{BCFTOOLS} query -f '%POS\\\\n' {os.path.join(WORK_DIR, 'genotypes.bcf')} > {os.path.join(WORK_DIR, 'site.positions')}\")\n",
    "\n",
    "\n",
    "positions = pd.read_csv(os.path.join(WORK_DIR, 'site.positions'), header = None)[0].values\n",
    "\n",
    "\n",
    "contig.recombination_map.get_rates()\n",
    "\n",
    "\n",
    "rate = contig.recombination_map.get_rates()[0] * 100 # convert to cM\n",
    "cM_pos = [pos*rate for pos in positions]\n",
    "\n",
    "with open(os.path.join(WORK_DIR, 'genetic_map.txt') , 'w') as OUTFILE:\n",
    "    for i in range(len(positions)):\n",
    "        OUTFILE.write(f\"22\\t{positions[i]}\\t{cM_pos[i]:0.6f}\\n\")\n",
    "os.system(f\"head {os.path.join(WORK_DIR, 'genetic_map.txt')}\")\n",
    "os.system(f\"tail {os.path.join(WORK_DIR, 'genetic_map.txt')}\")\n",
    "\n",
    "with open(os.path.join(WORK_DIR, 'sample_map.txt') , 'w') as OUTFILE:\n",
    "    for ind_string in ind_labels[:nref]:\n",
    "        pop = ind_string.split('-')[0]\n",
    "        OUTFILE.write(f'{ind_string}\\t{pop}\\n')\n",
    "os.system(f\"head {os.path.join(WORK_DIR, 'sample_map.txt')}\")\n",
    "os.system(f\"tail {os.path.join(WORK_DIR, 'sample_map.txt')}\")\n",
    "\n",
    "os.system(f\"{BCFTOOLS} query -f '%POS %REF %ALT\\\\n' {os.path.join(WORK_DIR, 'genotypes.bcf')} > {os.path.join(WORK_DIR, 'site.alleles')}\")\n",
    "\n",
    "site_alleles = pd.read_csv(os.path.join(WORK_DIR, 'site.alleles'), comment='#', sep = ' ', header = None)\n",
    "site_alleles.columns = ['pos', 'ref', 'alt']\n",
    "site_REF = site_alleles['ref'].tolist()\n",
    "site_ALT = site_alleles['alt'].tolist()\n",
    "\n",
    "df = pd.DataFrame(la_mat)\n",
    "#df.columns = [f'hap_{s}' for s in hap_labels]\n",
    "df.insert(loc = 0, column= 'CHROM', value = 22)\n",
    "df.insert(loc = 1, column= 'POS', value = positions)\n",
    "df.insert(loc = 2, column= 'ID', value = [f'site_{x.id}' for x in export_ts.sites()])\n",
    "df.insert(loc = 3, column= 'REF', value = site_REF)\n",
    "df.insert(loc = 4, column= 'ALT', value = site_ALT)\n",
    "df.insert(loc = 5, column= 'QUAL', value = '.')\n",
    "df.insert(loc = 6, column= 'FILTER', value = 'PASS')\n",
    "df.insert(loc = 7, column= 'INFO', value = '.')\n",
    "df.insert(loc = 8, column= 'FORMAT', value = 'LA')\n",
    "\n",
    "\n",
    "metadata = df.iloc[:,0:9].copy()\n",
    "metadata\n",
    "\n",
    "\n",
    "def vcfheader(ts, target_pop):\n",
    "    from datetime import date\n",
    "    fileformat = '##fileformat=VCFv4.2\\n'\n",
    "    fileDate = f'##filedate={date.today().ctime()}\\n'\n",
    "    FILTER = '##FILTER=<ID=PASS,Description=\"All filters passed\">\\n'\n",
    "    contig = f'##contig=<ID=22,length={int(np.ceil(ts.get_sequence_length()))}>\\n'\n",
    "    FORMAT = '##FORMAT=<ID=LA,Number=1,Type=String,Description=\"Local ancestry\">\\n'\n",
    "\n",
    "    pop_of_sample = dict(zip(range(len(ts.tables.nodes)), ts.tables.nodes.population))\n",
    "\n",
    "    nind_in_pop = collections.defaultdict(int)\n",
    "    pops = [pop_of_sample[i] for i in export_ts.samples(target_pop)]\n",
    "    for p in pops:\n",
    "        nind_in_pop[p] += 1\n",
    "    for p in nind_in_pop:\n",
    "        nind_in_pop[p] = int(nind_in_pop[p]/2)\n",
    "\n",
    "    ind_labels = []\n",
    "    for p in nind_in_pop:\n",
    "        for i in range(1, nind_in_pop[p]+1):\n",
    "            ind_labels.append(f'pop_{p}-ind_{i:05}')\n",
    "\n",
    "    header = fileformat+fileDate+FILTER+contig+FORMAT\n",
    "    header = header +'\\t'.join(['#CHROM','POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT'])\n",
    "    header = header + '\\t' + '\\t'.join(ind_labels) + '\\n'\n",
    "    return(header)\n",
    "\n",
    "header = vcfheader(export_ts, target_pop = admixed_pop_idx)\n",
    "\n",
    "\n",
    "\n",
    "read_fd, write_fd = os.pipe()\n",
    "write_pipe = os.fdopen(write_fd, \"w\")\n",
    "with open(os.path.join(WORK_DIR, 'la_true.bcf'), \"w\") as bcf_file:\n",
    "    proc = subprocess.Popen(\n",
    "        [f\"{BCFTOOLS}\", \"view\", \"-O\", \"b\"], stdin=read_fd, stdout=bcf_file\n",
    "    )\n",
    "    #print(header, file=write_pipe)\n",
    "    for line in header:\n",
    "        write_pipe.write(line)\n",
    "\n",
    "    for i in range(la_mat.shape[0]):\n",
    "        line = '\\t'.join(metadata.iloc[i].astype(str).tolist() + list(np.char.add(np.char.add(la_mat[i, ::2].astype(str),  '|' ), la_mat[i, 1::2].astype(str)))) + '\\n'\n",
    "        write_pipe.write(line)\n",
    "\n",
    "    write_pipe.close()\n",
    "    os.close(read_fd)\n",
    "    proc.wait()\n",
    "    if proc.returncode != 0:\n",
    "        raise RuntimeError(\"bcftools failed with status:\", proc.returncode)\n",
    "\n",
    "\n",
    "os.system(f\"{BCFTOOLS} view --output {os.path.join(WORK_DIR, 'la_true.vcf.gz')} --output-type z {os.path.join(WORK_DIR, 'la_true.bcf')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fb4ea61",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
